test/integration/acceptance/fake_peer_example_test.cpp

```diff
@@ -246,115 +253,123 @@ TEST_F(FakePeerExampleFixture, SynchronizeTheRightVersionOfForkedLedger) {
-   */
-  struct CustomBehaviour : public fake_peer::HonestBehaviour {
-    CustomBehaviour(const interface::types::HashType &tx_hash,
-                    std::atomic_flag &got_proposal_from_main_peer)
-        : tx_hash_(tx_hash),
-          got_proposal_from_main_peer_(got_proposal_from_main_peer) {}
-
-    void processYacMessage(
-        std::shared_ptr<const fake_peer::YacMessage> message) override {
-      const auto proposal_from_main_peer = getFakePeer().sendProposalRequest(
-          message->front().hash.vote_round, kProposalWaitingTime);
-      if (proposal_from_main_peer
-          and std::any_of(proposal_from_main_peer->transactions().begin(),
-                          proposal_from_main_peer->transactions().end(),
-                          [this](const auto &tx) {
-                            return tx.reducedHash() == tx_hash_;
-                          })) {
-        got_proposal_from_main_peer_.test_and_set(std::memory_order_relaxed);
-      }
-      HonestBehaviour::processYacMessage(message);
-    }
-
-    const interface::types::HashType &tx_hash_;
-    std::atomic_flag &got_proposal_from_main_peer_;
-  };
-
   // Create the tx:
   const auto tx = complete(
       baseTx(kAdminId).transferAsset(kAdminId, kUserId, kAssetId, "tx1", "1.0"),
       kAdminKeypair);
-  const auto hash = tx.reducedHash();
-
-  std::atomic_flag got_proposal_from_main_peer = ATOMIC_FLAG_INIT;
-
-  auto &itf = prepareState(1);
-  fake_peers_.front()->setBehaviour(
-      std::make_shared<CustomBehaviour>(hash, got_proposal_from_main_peer));
 
-  // watch the proposal requests to fake peer
-  bool got_proposal_from_fake_peer = false;
-  fake_peers_.front()->getProposalRequestsObservable().subscribe(
-      [&got_proposal_from_fake_peer](const auto &round) {
-        got_proposal_from_fake_peer = true;
+  auto proposal_storage = std::make_shared<fake_peer::ProposalStorage>();
+
+  createFakePeers(1);
+  fake_peers_.front()->setProposalStorage(proposal_storage);
+
+  auto &itf = prepareState();
+
+  // provide the proposal
+  proposal_storage->setDefaultProvider(
+      [tx](
+          const auto &round) -> fake_peer::OrderingProposalRequestResult {
+        shared_model::proto::ProtoProposalFactory<
+            validation::AlwaysValidValidator>
+            valid_factory;
+        std::shared_ptr<const shared_model::proto::Proposal> proposal;
+        valid_factory
+            .createProposal(round.block_round,
+                            iroha::time::now(),
+                            std::vector<shared_model::proto::Transaction>({tx}))
+            .match(
+                [&proposal](iroha::expected::Value<std::unique_ptr<
+                                shared_model::interface::Proposal>> &v) {
+                  // casting std::shared_ptr<shared_model::interface::Proposal>
+                  // to std::shared_ptr<const shared_model::proto::Proposal>
+                  proposal = std::const_pointer_cast<
```

Comment:
```suggestion
                  proposal = std::static_pointer_cast<
```
Because static cast allows to add `const`.